### ISSUE TITLE
Plugins!

### ISSUE THREAD

**Issue by terreng** [terreng/simple-web-server#97]
Plugins would allow anyone to create additional features or customize the behavior of the app. We could have a list of additional plugins on the website. I think a plugins system would strike a nice balance between keeping the app simple, but also letting people take it further if they want additional functionality.

Plugins would consist of a config file and some javascript. The config file would have details like the plugin name and version, and an array of additional options (that could be checkboxes, textboxes, or number inputs). In the UI, it would add an extra options section for each plugin with these options. Then in the javascript, you could write a custom script that can apply custom logic to requests, like adding headers or turning it into a proxy. This would work like the custom scripts/custom request handler features you've already made - it would have access to `req` and `res`, as well as the values of all the server options.

Users could download plugins - possibly as a zip of multiple files or perhaps as one combined file with a custom file extension - and then add them into the app from the UI or by placing them in a directory within the app's data folder like `/plugins`.

We'd want to make sure that it's easy to write basic plugins for things like adding additional headers. Ideally, this should be possible with only a few lines of code.

Alongside the introduction of plugins, I'd like you to create the first official plugin which adds the proxy feature.

Let me know what you think!

Here are some prototype/example screenshots of what it might look like:

<img width="363" alt="Screen Shot 2022-06-26 at 7 45 48 PM" src="https://user-images.githubusercontent.com/11605395/175850901-3b3120d9-640a-42f4-89f6-52bfb3c2de4a.png">

<img width="446" alt="Screen Shot 2022-06-26 at 7 49 36 PM" src="https://user-images.githubusercontent.com/11605395/175850915-99c1f37f-f0a6-44bf-905d-224b9d94cffe.png">

**Comment by ethanaobrien** [comment]
That sounds like a really good idea!
I think the snippet of json you have is a really good idea on how to store the data
I have some questions about the feature
1. Would there be a place in the app you can browse and download these plugins or would you need to do so manually?
2. Security. We cant just validate a key if we are transferring files over the internet. Something like how npm does it we need to find a way to validate the security of the javascript file before running it. This would be difficult for me. Do you know how I/we could do this?
3. Unhandled requests. How would we respond unhandled requests. They will be caught and logged (I added a listener for whenever anything throws an error so we at least have it)
4. With the proxy feature at least that would completely handle the request itself. I think we need a boolean value in the json that will tell the server whether or not to handle the request afterwards.
5. Ability to disable built-in options would be needed. Could we do this?
6. The proxy plugin at least needs to set up a websocket listener when the server object is initialized. I think there should be an option for a function to run on server startup that provides the server object.

I think this sounds like a really good idea!

**Comment by terreng** [comment]
Great questions! And I'm glad you like the idea.

> 1. Would there be a place in the app you can browse and download these plugins or would you need to do so manually?

I'm not sure about adding a "plugin store" type of thing into the app, because I'm not sure the best way to handle plugins that other people make, and I think we might run into issues with app store approval if we do that. I think it would be easiest to have them downloadable on the website, and easy to drag and drop into the program. But thanks for the idea, I'll keep thinking about it.

> 2. Security. We cant just validate a key if we are transferring files over the internet. Something like how npm does it we need to find a way to validate the security of the javascript file before running it. This would be difficult for me. Do you know how I/we could do this?

Do you just mean making sure that the file wasn't modified during download?

> 3. Unhandled requests. How would we respond unhandled requests. They will be caught and logged (I added a listener for whenever anything throws an error so we at least have it)

What do you mean? Do you mean requests that a plugin just doesn't "return" for? Seeing as the program would be executing javascript in the plugin files, there's only so much you can do to keep it on the rails. A broken plugin could make a server fail to start, hang, or maybe even crash the program. I don't see this as an issue though, because people should only be using plugins that work and aren't malicious.

> 5. Ability to disable built-in options would be needed. Could we do this?

Do you mean forcing an option to have a specific value, and preventing it from being changed in the UI? If so, what's an example use case? I was thinking a plugin could just ignore an option if it wants to.

> 4. With the proxy feature at least that would completely handle the request itself. I think we need a boolean value in the json that will tell the server whether or not to handle the request afterwards.
> 6. The proxy plugin at least needs to set up a websocket listener when the server object is initialized. I think there should be an option for a function to run on server startup that provides the server object.

The config file could specify multiple "hooks", like `before_start`, `before_request`, `before_response`, etc.

**Comment by ethanaobrien** [comment]
>I'm not sure about adding a "plugin store" type of thing into the app, because I'm not sure the best way to handle plugins that other people make, and I think we might run into issues with app store approval if we do that. I think it would be easiest to have them downloadable on the website, and easy to drag and drop into the program. But thanks for the idea, I'll keep thinking about it.

Makes sense.

>Do you just mean making sure that the file wasn't modified during download?

Yes, to make sure the request wasnt intercepted. This is more for if we decide to do the plugin store thing.

>What do you mean? Do you mean requests that a plugin just doesn't "return" for? Seeing as the program would be executing javascript in the plugin files, there's only so much you can do to keep it on the rails. A broken plugin could make a server fail to start, hang, or maybe even crash the program. I don't see this as an issue though, because people should only be using plugins that work and aren't malicious.

Im saying that what if an error is thrown processing the request. Would we be able to respond to it?

>Do you mean forcing an option to have a specific value, and preventing it from being changed in the UI? If so, what's an example use case? I was thinking a plugin could just ignore an option if it wants to.

The proxy option wouldnt need options like htaccess or choose path or any of the upload/deleting options and it seems really bulky to leave un-needed options in the UI.

>The config file could specify multiple "hooks", like before_start, before_request, before_response, etc.

Sounds good.

**Comment by terreng** [comment]
> Im saying that what if an error is thrown processing the request. Would we be able to respond to it?

I guess this depends on how exactly we implement it. I believe server side javascript is currently implemented using eval, right? Is it possible to execute a function from a file, and give it access to context like res, req, and opts?

It would be great if you could look in to how plugins would be implemented, and what hooks it makes sense to have.

> The proxy option wouldnt need options like htaccess or choose path or any of the upload/deleting options and it seems really bulky to leave un-needed options in the UI.

I see. Let me think about this one.

**Comment by ethanaobrien** [comment]
>I believe server side javascript is currently implemented using eval, right?

Correct

>Is it possible to execute a function from a file, and give it access to context like res, req, and opts?

I think the only thing we could really do is what I already have, to surround it with a try/catch block, although it wouldnt be caught if it went into a callback function

>It would be great if you could look in to how plugins would be implemented, and what hooks it makes sense to have.

Will do. We would probably need at least the following. Let me know if I am missing any

Functions:
`onStart`: will be called when the server is created. Will include the server object
`onRequest`: will be called when an incoming request is received. The arguments will be the `req` and `res` objects along with a `preventDefault` function (for when the plugin is made to handle the entire request, will cancel the default processing of reading and sending the file.)

JSON args:
id, options, version, main (as you showed in the screenshot).
and probably (If you agree) a json array of the options to hide.

I think this is all. Let me know if there are any things I am missing.


>I see. Let me think about this one.

Sounds good

**Comment by terreng** [comment]
Good idea with using a preventDefault. Using a try/catch sounds like the best approach. 

As for disabling options, I'm thinking about how users would active plugins on servers. I was thinking the plugin's options would show up on every server, but maybe instead the user would specifically enable the plugin on each server. If we did it this way, then it would make more sense for certain options to disappear when you do that. I'll keep thinking about how best to implement it.

**Comment by ethanaobrien** [comment]
Sounds good!

Another good idea is maybe to also have an option in the json file that tells the app that no other plugins can be enabled while this plugin is in use, because with complicated plugins (like the proxy plugin) additional plugins may break the plugin

**Comment by ethanaobrien** [comment]
@terreng I have pushed the beta plugin code. Let me know what you think.

Here is a simple plugin that all it does is either override the response and sends something itself or it just adds a test header. You can change which in the `main.js` file

[test-plugin.zip](https://github.com/terreng/simple-web-server/files/9015773/test-plugin.zip)

**Comment by terreng** [comment]
Looks good! Thanks for the example plugin. Could you make a new draft pr from that branch that closes this issue?

Thanks for working to make the code more modular by moving bookmarks stuff to its own file! However, the change actually breaks the code because `mas_bookmarks` and `addToSecurityScopedBookmarks` are used in `index.js`. Do you think that you could fix that?

When it comes to options, I was thinking that they could be stored as `[pluginid].[optionname]` in the config file. So plugin options are stored together with built in options, but prefixed with the plugin's id. The plugin config file should have an `id` field, and that should be used to differentiate plugins internally. If you install a new plugin with the same id as an existing plugin, it replaces the existing one (like for updating plugins). So the id field should be unique per plugin, but the display name can vary.

**Comment by ethanaobrien** [comment]
>When it comes to options, I was thinking that they could be stored as [pluginid].[optionname] in the config file

I actually have a better idea. If the plugin value was an array this array could be filled with the plugin info along with the config of the plugin. I went ahead and did this because its much cleaner, let me know if you want to change it.

>the change actually breaks the code because mas_bookmarks and addToSecurityScopedBookmarks are used in index.js. Do you think that you could fix that?

Fixed.

I also changed the thing to use require (do we need to call a bookmark thing for require if its outside the app?)

Let me know how it looks so far.


Also, I updated the example plugin to go along with me using require to execute the main file. Heres the new example

[test-plugin.zip](https://github.com/terreng/simple-web-server/files/9024720/test-plugin.zip)

**Comment by ethanaobrien** [comment]
@terreng if I’m correct I finished all the backend stuff. Let me know if there’s anything more I need to do

**Comment by terreng** [comment]
Could you now turn the proxy feature into a plugin? This can be a separate repo that you own.

**Comment by ethanaobrien** [comment]
>Could you now turn the proxy feature into a plugin?

Will do, one question first.

What should the file that stores the information be called? I set it to `manifest.json` for testing but this can interfere if the plugin uses nodejs (like the proxy plugin). What should the json info file be named?

I also think we should add the ability to import zip files (so the user could download the zip and just import without having to unpack). Maybe [JSZip](https://www.npmjs.com/package/jszip) to unpack this to the plugins folder?


Also, thought I'd explain what all the plugin functions do.

`registerPlugins(data)`: for when the server is started, attaches any plugins to the server. 
`data`: the `this_server` object
Returns the needed callbacks along with other data required to operate the function (stored in `this_server.plugin`)

`importPlugin(path, config)`: to import a plugin to the server
`path`: The path to the plugin folder.
`config`: The `serverconfig.plugin` value.
Returns the new value to be set to `serverconfig.plugin` and saved.

`removePlugin(id, serverConfig, config)`: removes the plugin.
`id`: Id of plugin to remove
`serverConfig`: The array of all the servers (used to check if the server needs to be deleted from the plugins folder).
`config`: The `serverconfig.plugin` object of the server we are removing the plugin from.
Returns the new `serverconfig.plugin` object to be set and saved.

**Comment by terreng** [comment]
How about `plugin.json`?

Yeah, if the user adds a zip file, we can unzip it and add it to the plugins directory.

Here's what I'm thinking for the UI: A place to view the list of installed plugins, add new ones, and remove plugins. On each server, an option to enable/disable each plugin on that server. I'm thinking that each plugin has an options section on every server, with a switch on it to enable/disable the plugin for that server.

I'm a little unclear on how it is implemented now. Are plugins added per server, or to all the servers?

Plugins should be installed/uninstalled at the app level.

Once a plugin is installed, it should be enabled/disabled per server.

**Comment by ethanaobrien** [comment]
>How about `plugin.json`?

Got it.

>I'm a little unclear on how it is implemented now. Are plugins added per server, or to all the servers?

per server

>I'm a little unclear on how it is implemented now. Are plugins added per server, or to all the servers?

I currently have it per server (kind of) I should probably do what you suggested. I have more details in the reply below.

>Once a plugin is installed, it should be enabled/disabled per server.

So you're saying to have a separate way to delete the plugin entirely?
This wouldn't be too hard. I would just have the remove plugin function instead delete it from the plugin array and create a new function to remove it from the app data folder (to delete it entirely). Also I would need, in the import function, to replace the path with the id of the plugin to register, and have a separate function to store the plugin in the app data folder. This wouldn't be too hard.

**Comment by terreng** [comment]
Here's what I'm thinking. Let me know if you can do this or if you think we should do it differently:

**Installing plugins**

Add a folder to the `/plugins` directory of the app data folder. The plugin folder name doesn't matter. Inside it must have a plugin.json file to be recognized.

Example:
```
- plugins
   - proxy
       - plugin.json
       - script.js
```

In the app, installing a plugin through the UI adds a folder to `/plugins`. Uninstalling a plugin deletes the corresponding folder. The app determines the list of available plugins by scanning the plugins directory.

**Enabling plugins per server**

Let's have a new server option called `plugins`. It's an object, and by default it's empty `{}`.

A plugin is enabled on a server if its id is a key of the object. For example, this value of the `plugins` option would indicate that the `proxy` plugin is enabled on the server:

```
{
  "proxy": {}
}
```

Inside the inner object is where the proxy plugin stores the values for its options.


As you can see, installing and uninstalling plugins does not change config.json in this model. If a plugin is uninstalled, its options stay within the servers, but they are ignored because no plugin exists with that id.

**Comment by ethanaobrien** [comment]
I see, so it’s very similar to what I currently have. I’ll get that done

> Would most plugins need this option?

Any plugins that completely change the way the server behaves. Most, I’m not sure.

> I pictured most plugins being things that don't completely change the server.

I think they could be both ways, in my view most of the things that don’t completely change the server can be done with htaccess so wouldn’t the more complicated things that change the way the server behaves be used here?

>For example, would it be possible to implement htaccess files using only a plugin?

Yes, this would be completely possible. You can do anything with the plugins (like re-write the entire handler (that’s similar to what web server chrome was for if you look at the readme for the project.)) what do you think?

**Comment by terreng** [comment]
> what do you think?

Do you want to make the htaccess feature a plugin?

Also, is this guide for custom scripts, or something else? https://simplewebserver.org/docs/custom%20request%20handler.html#step-1-create-the-request

I'll add the UI for installing and uninstalling plugins, and changing their options on servers. Could you add the code that scans the `/plugins` directory and populates a global `plugins` variable with an array of their configs. Do you think we should make it so that the name of the plugin directory must match the plugin's id? This could be done at the time of install.

**Comment by ethanaobrien** [comment]
> Do you want to make the htaccess feature a plugin?

I’d really rather not, the htaccess feature is incredibly embedded and would be a pain to remove.

> Also, is this guide for custom scripts, or something else?

Yes, that’s is the custom scripts guide.

> Could you add the code that scans the /plugins directory and populates a global plugins variable with an array of their configs.

Will do

> Do you think we should make it so that the name of the plugin directory must match the plugin's id?

it doesn’t really matter, if you would like

**Comment by terreng** [comment]
> it doesn’t really matter, if you would like

If we don't, then the global array of plugin configs needs to add a directory name attribute to each one.

**Comment by ethanaobrien** [comment]
Ok, yeah I can make it the ID.

**Comment by ethanaobrien** [comment]
@terreng The proxy plugin is ready! You can view it here - https://github.com/ethanaobrien/web-proxy (note, the web proxy `index.js` file is a standalone version of the web proxy).

Let me know what you think!

**Comment by terreng** [comment]
I'll add plugins to the UI next week probably.

Also, let's add a few more optional arguments for plugin.json:
- author
- homepage
- license
- description

**Comment by terreng** [comment]
It's probably going to be another month before I work on this. Just wanted to give you an update.

**Comment by ethanaobrien** [comment]
All good!

**Comment by terreng** [comment]
I've had a great experience using the program over the last few months. I'm really happy with what we've built together. 183 people have downloaded it from the Microsoft Store, 90 from the Mac App Store, and approximately 20-40 downloads from GitHub (I don't remember how many of them were me). We have a few 5 star ratings on the Microsoft Store, and one review that says "Great! Works well. Easy to use."

**Comment by ethanaobrien** [comment]
Awesome! It is really cool with what we built.

I was thinking, on some computers something like this is a little, heavy (on resources), what do you think about, instead of completely translating the whole app, having like a “mini version” with only the basics written in another language. What do you think about this. I think C# would be a choice we have, but I haven’t ever actually tried it on macOS and Linux and I don’t know if it requires downloading dot net or anything.

I have a simple C# server you can take a look at (it’s pretty close to being done, there are a few final bugs), and if you have any interest in doing it, if you go back in the commit history it was originally written in plain C. I don’t recommend using plain C though, it took me a single day to rewrite most of the features into C# that  took me months in regular C (and it’s not easily cross platform). Here’s the project: https://github.com/ethanaobrien/C-server

**Comment by terreng** [comment]
I agree that the app isn't particularly resource efficient. I think the most notable part of this is the large file size. I like the idea of rewriting the web server code in something closer to the metal.

I've been keeping an eye on a cool new technology called Tauri, which has officially launched version 1.0. Tuari works kinda like Electron, but instead of bundling Chrome and Node, it uses the system webview (WebKit/Safari on macOS, Chromium on Windows 11) and Rust. Rust is nearly as fast as C, and has a really good ecosystem of packages (called crates).

What's great about switching to Tauri is that (pretty much) none of the UI would need to change! With Tauri, the app would be super memory efficient because only the Rust code would be running when the program isn't visible, and the package size would be WAY smaller because it wouldn't need to bundle Chromium. 

I know I've mentioned Rust before. I think Rust would be a better choice than C/C# for this project because Tauri is built around it. I haven't done much in Rust, but I don't imagine it's any harder than C#.

What do you think about making a new branch and building a basic web server using Rust? It doesn't have to support all the features of the app, just more as a proof of concept / mini version. If that works out, then maybe we rewrite the rest of the web server code to use Rust? What do you think?

**Comment by ethanaobrien** [comment]
I think thats a good idea. Would you mind sharing with me a simple demo project and get me the link to the compiler and the docs (if any).

I dont know about re-writing the htaccess and plugins and server side functions. I think we should have a lite mode (like the one I described before) that on first launch it asks if you want to use the lite mode or the electron version and you can switch between the two (would require restart which is pretty obvious) but I think this is the best approach. What do you think?

Thanks!

**Comment by terreng** [comment]
You're right that the server side javascript stuff and plugins wouldn't work unless we found a way to execute javascript. The Tauri version of the app would have to give up those features, but I suppose people could still download the Electron version if they want those features.

I haven't messed around much with Tauri yet. [Maybe follow the demo app tutorial](https://tauri.app/v1/guides/getting-started/setup/html-css-js) to get a basic project, and see if you can figure out how to get a basic web server up and running. I can try getting a basic Tauri project started too, in a little while.

**Comment by ethanaobrien** [comment]
Hello, I finally got a chance to sit down and look into building a simple rust server today and it has proven to be very difficult. When compiling I kept getting an openssl utils compile error and everything I found online was all for Linux systems. Because of this I'm not sure how easy it would be to make it cross platform. What do you think @terreng ?

**Comment by terreng** [comment]
Do you want to share your code and the error with me? I have a feeling that this won't be too difficult to sort out. If you want, I can have a go at building a basic Rust prototype.

Very cool that you've written a web server in C#! However, the reason we should use Rust is because it's what Tauri primarily supports, and we need Tauri in order to copy over the existing user interface. 

I don't think we need to build native versions for each platform. Using Rust and Tauri will already be significantly more efficient than Nodejs and Electron. Plus, we can write once and support Windows, macOS, and Linux. It looks like Tauri has mobile support coming soon too!

I really do think that Rust and Tauri is the way forward.

**Comment by ethanaobrien** [comment]
Here is what I have so far: 
[rust server.zip](https://github.com/terreng/simple-web-server/files/9691448/rust.server.zip)

The error I get is `failed to run custom build command for 'openssl-sys-extras v0.7.14'` when compiling on windows

We might want to keep other languages as a possibility just in case something doesnt work, but hopefully Tauri works for what we need it to

**Comment by terreng** [comment]
@ethanaobrien I'm planning to work on building the UI for plugins in December - January. I'm also planning to add a dark mode UI #14, the ability to reorder servers #62, and help icons / documentation #66. This will also close this issue as well as #9. It'll be version 1.2.

Just to confirm, you've tested the [proxy plugin](https://github.com/ethanaobrien/web-proxy) and everything works, right? Is everything in that repo part of the plugin?

**Comment by terreng** [comment]
Would be great if we could also get fixes for #101 and #107 into version 1.2.

**Comment by ethanaobrien** [comment]
>Just to confirm, you've tested the [proxy plugin](https://github.com/ethanaobrien/web-proxy) and everything works, right? Is everything in that repo part of the plugin?

The only thing I have left to do is the ability to import zips

>Would be great if we could also get fixes for https://github.com/terreng/simple-web-server/issues/101 and https://github.com/terreng/simple-web-server/issues/107 into version 1.2.

I should be able to look at these this week. Sorry for the delay

**Comment by terreng** [comment]
> The only thing I have left to do is the ability to import zips

You mean for the program? Or for the plugin?

**Comment by ethanaobrien** [comment]
To import plugins from a zip file. Sorry for the misunderstanding

**Comment by terreng** [comment]
This would just be unzipping the file and copying its contents to the plugins directory, right? I'm assuming we'll trigger that if you pick a zip file when adding a plugin in the UI.

**Comment by ethanaobrien** [comment]
Correct. On both statements

**Comment by terreng** [comment]
Okay sounds good. Could you make a PR that adds a function which takes in the path to a zip file and takes care of adding it?

Also, could you add a function that scans the plugin directory and returns an array of all the plugin configs?

As for the proxy plugin, is everything in the repo just for the plugin? It looked like a lot of files, so I wanted to check. If not, could you make a new repo just for the plugin? And could you add a little bit of documentation to the README explaining that it's a plugin for Simple Web Server and explaining what the arguments do?

**Comment by ethanaobrien** [comment]
>Could you make a PR that adds a function which takes in the path to a zip file and takes care of adding it?

I can do that

>Also, could you add a function that scans the plugin directory and returns an array of all the plugin configs?

I already have `getInstalledPlugins` which returns an array of the installed plugins and then `getPluginInfo` which will get you the manifest. Do you want me to combine these?

>As for the proxy plugin, is everything in the repo just for the plugin?

Its actually a standalone app, with plugin compatibility. Yes everything in there is just for the proxy server. If you'd like me to make another repo, I could do so, but that would basically be the same thing in a different place.

>And could you add a little bit of documentation to the README explaining that it's a plugin for Simple Web Server and explaining what the arguments do?

Can do

**Comment by terreng** [comment]
> I already have `getInstalledPlugins` which returns an array of the installed plugins and then `getPluginInfo` which will get you the manifest. Do you want me to combine these?

Oh perfect! No need to combine, that's already great.


I think it would be better if you made it a separate repo with only the plugin, and included documentation.

**Comment by terreng** [comment]
@ethanaobrien Could you make the function that adds plugins from a ZIP work with directories too? I'm making it so that you can pick either a ZIP file or a directory from the UI. It would be great if, given the path, the function could check if it's a zip or not and do the right thing.

Also, I noticed that `getInstalledPlugins` is called once when the program starts up. Could you make the function that adds a new plugin also refresh the list of plugins so that you don't have to restart the program to use them?

Finally, could you add a function that removes a plugin given its id? And have that refresh the list of plugins too so that it gets disabled immediately.

**Comment by terreng** [comment]
So far on #114 I've added dark mode, the ability to reorder servers, included documentation, and made it sync config changes in real time (so the UI updates when you edit config.json while the program is open).

**Comment by ethanaobrien** [comment]
> Could you make the function that adds plugins from a ZIP work with directories too?

In theory, the current input function should be able to handle both. Does it not work?


> Also, I noticed that getInstalledPlugins is called once when the program starts up. Could you make the function that adds a new plugin also refresh the list of plugins so that you don't have to restart the program to use them?

Yeah, I can do that if you'd like. The getInstalledPlugins is actually called in the index.js file, not the plugin function itself and it just returns the set plugins, it itself doesn't set the plugins.


> Finally, could you add a function that removes a plugin given its id? And have that refresh the list of plugins too so that it gets disabled immediately.

The `removePlugin` function will remove a plug-in based on its id, it should at least. I can also add on that it will update the global plugins object if that's what you'd like

**Comment by terreng** [comment]
Oh, I missed that you already had written `importPlugin` and `removePlugin`. My bad.

It would be great if you could update them however needed so that adding and removing plugins takes effect immediately. 

Also, could you use [`fs.watch`](https://nodejs.org/docs/latest/api/fs.html#fswatchfilename-options-listener) to watch for changes to the plugin directory, and reload the plugins when anything changes? This would enable someone to edit the plugin's javascript and have the changes take effect in real time, which would be cool.

**Comment by ethanaobrien** [comment]
> Oh, I missed that you already had written importPlugin and removePlugin. My bad.

It's all good

> It would be great if you could update them however needed so that adding and removing plugins takes effect immediately.

When a plug-in is removed/added, the server the plug-in is on will need to be restarted. I'll add a function that it calls when these things happen that will set the global plugins variable.

> Also, could you use fs.watch to watch for changes to the plugin directory, and reload the plugins when anything changes? This would enable someone to edit the plugin's javascript and have the changes take effect in real time, which would be cool.

We can do this, although if the main plug-in config file was modified, the server that the plug-in is attached to would need to restart

**Comment by ethanaobrien** [comment]
Also, how will `fs.watch` work with macos security bookmarks?

**Comment by terreng** [comment]
> Could you write the part that restarts only the needed servers. I saw you managed to do this when settings were modified

Sure. I assume that it should restart any servers that the plugin is enabled on? If so I'll write a function like `restartServersWithPlugin(pluginid)`. Does that work?

> Also, how will `fs.watch` work with macos security bookmarks?

The app always has permission to access the contents of its own app data directory, so security scoped bookmarks don't matter. This will only become a problem if we want to use `fs.watch` on directories outside, like a folder the user has selected.

**Comment by ethanaobrien** [comment]
> Sure. I assume that it should restart any servers that the plugin is enabled on? If so I'll write a function like restartServersWithPlugin(pluginid). Does that work?

Yes, this would work.

> The app always has permission to access the contents of its own app data directory, so security scoped bookmarks don't matter. This will only become a problem if we want to use fs.watch on directories outside, like a folder the user has selected.

Ok good. Are we going to include that for development purposes you can actively edit in the plugins folder?

**Comment by terreng** [comment]
> Ok good. Are we going to include in the documentation that for development purposes you can actively edit in the plugins folder?

Sure. Speaking of documentation, we'll need to document the plugin.json format and javascript handlers you can write. Would you like to take on writing the documentation for plugins? Or we can do it together.

**Comment by terreng** [comment]
I added a function `restartServersWithPlugins(pluginids)` to the main branch. It takes in an array of plugin ids.

**Comment by ethanaobrien** [comment]
> Would you like to take on writing the documentation for plugins? Or we can do it together.

I'd like to try to write it, and then have you revise as needed, if that's ok with you


I may not be able to work on any of this until next week, I've been really busy with work. Sorry about that

**Comment by terreng** [comment]
> I'd like to try to write it, and then have you revise as needed, if that's ok with you

Sounds good.

> I may not be able to work on any of this until next week, I've been really busy with work. Sorry about that

No worries!

**Comment by terreng** [comment]
@ethanaobrien I've made some changes to the plugin code.

- I added a function `validatePluginManifest(manifest)` which checks to make sure that the manifest is all valid, including the options, to make sure it doesn't crash the UI. I also added this check before plugins are loaded in.
- I removed the `activate` and `deActivate` functions because we won't need them.
- I added a check to make sure that the folder name of plugins matches their id when loading them in. It looks like we're assuming everywhere that the ID of a plugin matches the name of the folder it's in, so we should enforce that wherever possible.
- In `index.js`, I renamed the global variables so that `plugin` refers to the import and `plugins` refers to the list of plugins.

All of these changes are on the `plugins-darkmode-reorder` branch, so you should probably work from that instead of `main`.

I came across an error that will only happen on macOS, which is related to the hidden `.DS_Store` file that macOS uses:

```
could not import plugin .DS_Store Error: Entered path is not directory
    at new FileSystem (/Users/terren/Documents/GitHub/web-server/WSC/FileSystem.js:223:19)
    at getPluginInfo (/Users/terren/Documents/GitHub/web-server/plugin.js:91:14)
    at Object.getInstalledPlugins (/Users/terren/Documents/GitHub/web-server/plugin.js:229:30)
    at Object.<anonymous> (/Users/terren/Documents/GitHub/web-server/index.js:28:25)
```

Could you make sure to ignore dot files in the plugins directory instead of erroring?

**Comment by terreng** [comment]
I got started writing documentation for plugins. I split it into three parts: Introduction to plugins, Plugin manifest file, and Plugin script. It would be great if you could write the Plugin script page to explain how scripts work. I wrote down some ideas you should make sure to include. The file is located at `website/src/docs/plugin script.md`.

**Comment by terreng** [comment]
@ethanaobrien I went ahead and implemented live reloading of plugins using `fs.watch` and the `restartServersWithPlugins` function I wrote.

The plugins UI is all done. And plugins overall are pretty much all done.

I think the only thing left for you to do is fix the `.DS_Store` issue I mentioned above, and write some documentation. 

Oh and also, make sure to clean up / make a new repo for the Proxy plugin, as we talked about. And two changes you should make to [`plugin.json`](https://github.com/ethanaobrien/web-proxy/blob/main/plugin.json):
- Because `forceSite` is a string, the value for `default` should be an empty string instead of false. The plugin manifest validation I wrote will reject this unless you fix it.
- For each option, you can add a `description` field to explain what it does, which will show up in the UI as a question mark you can click for more info.

**Comment by terreng** [comment]
Please go ahead and test everything on the `plugins-darkmode-reorder` branch and let me know if you find any bugs!

**Comment by ethanaobrien** [comment]
>I went ahead and implemented live reloading of plugins using fs.watch and the restartServersWithPlugins function I wrote.

Thank you, sorry for my inactivity. I've had a lot going on.

> I think the only thing left for you to do is fix the .DS_Store issue I mentioned

This should be easy. I'll do this as soon as I can

> and write some documentation

I'll do this when I can

> and also, make sure to clean up / make a new repo for the Proxy plugin

I actually determined that I'm going to keep it in the same repository, since the only difference would be that one has a plugin.json file and one doesn't (since the project itself is based as a plug-in). As for the proxy project itself, I do need to clean up and write the documentation



I should be able to test everything next week!

**Comment by terreng** [comment]
Sounds good! Once you're done with all that we should be ready to do a release.

**Comment by terreng** [comment]
One more issue I found. Importing plugins from ZIP files isn't working. Here's the error:
```
(node:28229) UnhandledPromiseRejectionWarning: TypeError: Cannot read properties of null (reading 'async')
    at /Users/terren/Documents/GitHub/web-server/plugin.js:157:70
```
This line:
```javascript
const manifest = JSON.parse(await zip.file('plugin.json').async("string"));
```
Could you take a look and fix it?

**Comment by ethanaobrien** [comment]
This is happening because it doesn't like how the zip is zipped. Everything at the moment needs to be in the base zip directory (when you open the zip file, the file should be there, no folders) there's no incredibly efficient way around this, although if no plugin.json was found we could check the first folder found. What do you think?

**Comment by terreng** [comment]
> Everything at the moment needs to be in the base zip directory (when you open the zip file, the file should be there, no folders) there's no incredibly efficient way around this, although if no plugin.json was found we could check the first folder found.

It would be great if we could make that work, because I can definitely imagine people making this mistake.

I'm still unable to import from a ZIP even when it's formatted correctly. There isn't anything in the terminal, but it's still erroring.

Here are the files I'm testing with. It works when I pick a folder with the files, but not when I pick the zip file.

[Archive.zip](https://github.com/terreng/simple-web-server/files/10300853/Archive.zip)

<details>
  <summary>plugin.json</summary>
  
```json
{
  "id": "my_example",
  "name": "Example\n\nPlugin",
  "script": "script.js",
  "options": [
    {
      "id": "my_checkbox",
      "name": "Example\n\ncheckbox",
      "description": "Enable if you like checkboxes.",
      "type": "bool",
      "default": false
    },
    {
      "id": "my_textbox",
      "name": "Example\n\ntextbox",
      "description": "Enter your favorite word.",
      "type": "string",
      "default": ""
    },
    {
      "id": "my_number",
      "name": "Example\n\nnumber input",
      "description": "Specify your favorite number.",
      "type": "number",
      "default": 50,
      "min": 0,
      "max": 100
    },
    {
      "id": "my_dropdown",
      "name": "Example\n\ndropdown",
      "description": "Choose your preferred color.",
      "type": "select",
      "default": "red",
      "choices": [
        { "id": "red", "name": "Re\n\nd" },
        { "id": "yellow", "name": "Ye\n\nllow" },
        { "id": "green", "name": "Gre\n\nen" },
        { "id": "blue", "name": "Blu\n\ne" }
      ]
    }
  ]
}
```
  
</details>

<details>
  <summary>script.js</summary>
  
```javascript
function onStart(server, options) {
    // Do nothing
}

function onRequest(req, res, options, preventDefault) {
    if (options.header == true) {
        res.setHeader('my-header', 'hello world');
    }
}

module.exports = {onStart, onRequest};
```
  
</details>

Happy Holidays by the way :)

**Comment by ethanaobrien** [comment]
Very sorry for the delay. I just pushed a change to the plugins-darkmode-reorder branch that should fix this issue. I also added some comments in the code of some bugs I found. I was also able to load the example plugin you find and run it (I had to modify it, because `options.header` returned undefined. I just set the header, no conditions, and it works)

![Screenshot 2022-12-31 11 33 21 PM](https://user-images.githubusercontent.com/77750390/210161982-e8fae6f2-dfaa-4e4b-82c6-c9adf7fbddc6.png)


Let me know if there are any more plugin problems!

Happy new years!

**Comment by terreng** [comment]
Thanks for the fix! I tried adding a plugin that was zipped improperly (folder inside the zip), and now I'm getting these errors:

```
copy /Users/terren/Library/Application Support/Simple Web Server/plugins/my_example
(node:43550) UnhandledPromiseRejectionWarning: TypeError: Cannot read properties of undefined (reading 'startsWith')
    at getZipFiles (/Users/terren/Documents/GitHub/web-server/plugin.js:101:28)
    at copyFolderRecursiveSyncFromZip (/Users/terren/Documents/GitHub/web-server/plugin.js:111:17)
    at /Users/terren/Documents/GitHub/web-server/plugin.js:156:19
could not import plugin .DS_Store Error: Entered path is not directory
    at new FileSystem (/Users/terren/Documents/GitHub/web-server/WSC/FileSystem.js:223:19)
    at getPluginInfo (/Users/terren/Documents/GitHub/web-server/plugin.js:91:14)
    at Object.getInstalledPlugins (/Users/terren/Documents/GitHub/web-server/plugin.js:245:30)
    at Timeout._onTimeout (/Users/terren/Documents/GitHub/web-server/index.js:210:102)
    at listOnTimeout (node:internal/timers:559:17)
    at process.processTimers (node:internal/timers:502:7)
Error stating "/Users/terren/Library/Application Support/Simple Web Server/plugins/my_example/plugin.json" Error: ENOENT: no such file or directory, stat '/Users/terren/Library/Application Support/Simple Web Server/plugins/my_example/plugin.json'
    at statSync (node:fs:1588:3)
    at t.statSync (node:electron/js2c/asar_bundle:2:4510)
    at getByPath.getFile (/Users/terren/Documents/GitHub/web-server/WSC/FileSystem.js:40:24)
    at FileSystem.getByPath (/Users/terren/Documents/GitHub/web-server/WSC/FileSystem.js:231:24)
    at getPluginInfo (/Users/terren/Documents/GitHub/web-server/plugin.js:92:34)
    at Object.getInstalledPlugins (/Users/terren/Documents/GitHub/web-server/plugin.js:245:30)
    at Timeout._onTimeout (/Users/terren/Documents/GitHub/web-server/index.js:210:102)
    at listOnTimeout (node:internal/timers:559:17)
    at process.processTimers (node:internal/timers:502:7) {
  errno: -2,
  syscall: 'stat',
  code: 'ENOENT',
  path: '/Users/terren/Library/Application Support/Simple Web Server/plugins/my_example/plugin.json'
}
could not import plugin my_example TypeError: fs.getByPath(...).text is not a function
    at getPluginInfo (/Users/terren/Documents/GitHub/web-server/plugin.js:92:60)
    at Object.getInstalledPlugins (/Users/terren/Documents/GitHub/web-server/plugin.js:245:30)
    at Timeout._onTimeout (/Users/terren/Documents/GitHub/web-server/index.js:210:102)
    at listOnTimeout (node:internal/timers:559:17)
    at process.processTimers (node:internal/timers:502:7)
```
No error in the UI. Here's the ZIP file I'm using. Would be great if you could try it:
[myplugin 2.zip](https://github.com/terreng/simple-web-server/files/10328398/myplugin.2.zip)

I'll work on fixing the bugs you identified. Thanks for pointing them out!

**Comment by terreng** [comment]
I fixed the file picker issue. As for fs.watch being unsupported, the best I can do is catch the error. It's totally not an essential feature, but unfortunately isn't available on Linux.

**Comment by ethanaobrien** [comment]
>As for fs.watch being unsupported, the best I can do is catch the error. It's totally not an essential feature, but unfortunately isn't available on Linux.

Yeah, It is a little disappointing.

I also fixed support for zips zipped in a way the app doesn't like. It will scan the entire zip for the `plugin.json` file, and copy whatever folder it is in

Let me know if you find anything else

**Comment by terreng** [comment]
Great, thanks! Could you also fix the .DS_Store issue? Just ignore anything in the plugins directory that isn't a folder.

**Comment by ethanaobrien** [comment]
I just pushed my changes. Everything should be good

**Comment by terreng** [comment]
Great! Unless you find anything else, I think the release is ready to go.

The only thing left is documentation. I wrote a getting started page and documented the plugin.json format. Please write some documentation for plugin scripts, and then I can edit it after you're done if you'd like. Also remember to update/add documentation to the proxy plugin repo.

Once you do that, we can release this version!

**Comment by ethanaobrien** [comment]
@terreng

I'll get started on the documentation when I can

The proxy plugin itself is done (not yet the documentation) here is a compressed zip loadable as a plugin.

[web-proxy.zip](https://github.com/terreng/simple-web-server/files/10333135/web-proxy.zip)

After looking at this, I do urge you to reconsider my proposed idea on an option to disable some options in plugins that basically overwrite the entire handler, as it may cause confusion why a plugin like a proxy is asking for directory or why most options do absolutely nothing. This might confuse the user, from my perspective, and I urge you to look at the plugin and think about it

**Comment by ethanaobrien** [comment]
@terreng Have we tried something like [this](https://stackoverflow.com/a/13705878)

**Comment by terreng** [comment]
Awesome, looks like that handle the nightmare of having a bunch of recursive watchers. Good find, I'll update the code to use that and it should work on all platforms.

**Comment by ethanaobrien** [comment]
Also, looking at the comments of that stack overflow post, make sure to set `awaitWriteFinish` to `true` to make sure that if a file is being copied or something, that we won't get spammed with change callbacks

**Comment by ethanaobrien** [comment]
Hey @terreng I just wrote a more simplistic plugin that I was wondering if you wanted to add to the list of plugins.

Heres the repo url: https://github.com/ethanaobrien/cloudflare-ip-whitelist

It basically acts as a whitelist for cloudflare ip addresses. Cloudflare has the proxy feature that proxys all the requests, and this plugin blocks any request not from cloudflare ips

**Comment by terreng** [comment]
@ethanaobrien Sure, I added it to the list.

You code is fine, but there are a few improvements you could make:
- Load the ipv4 and ipv6 IPs in parallel instead of one then the other
- Think about what should happen in the error case
- Think about what should happen for requests that hit before the IP ranges have loaded (I'd suggest that you enqueue these requests and have them wait)

By the way, do custom dependencies work in plugins?

**Comment by ethanaobrien** [comment]
@terreng 

>Load the ipv4 and ipv6 IPs in parallel instead of one then the other

I thought about this, the main issue was in the case of an error. I didn't want to delete the already cached ips, so I just loaded one after the other, since it would be the easiest way to see if they both succeeded.

>Think about what should happen in the error case

In case of an error, which is the same reason I didnt write it in parallel, it just keeps the old ips it already has.

>Think about what should happen for requests that hit before the IP ranges have loaded (I'd suggest that you enqueue these requests and have them wait)

Or itd probably be a good idea to just have what's on cloudflare's site in the program, since the changelog shows it doesn't get updated too often, just to do this as a backup.
